File: WoofIOS/SwiftPackages/Presentation/Home/Sources/PresentationHome/ViewModel/HomeViewModel.swift
import Combine
import DI
import PresentationCommon

@Factory
public final class HomeViewModel: StateProducingViewModel<HomeViewModel.State, Never> {
    init() {
        super.init(initialValue: .initial)
    }

    public func onTabSelect(_ tab: HomeTabUIModel) {
        setState(currentState.copy(selectedTab: tab))
    }

    public func onBackTap() {
        setState(currentState.copy(selectedTab: .browse))
    }
}

extension HomeViewModel {
    public struct State: Equatable {
        public let selectedTab: HomeTabUIModel
        public let tabs: [HomeTabUIModel]

        public var isBackHandled: Bool {
            selectedTab != .browse
        }

        public static let initial = State(selectedTab: .browse, tabs: HomeTabUIModel.allCases)

        public init(selectedTab: HomeTabUIModel, tabs: [HomeTabUIModel]) {
            self.selectedTab = selectedTab
            self.tabs = tabs
        }

        func copy(selectedTab: HomeTabUIModel) -> State {
            State(selectedTab: selectedTab, tabs: tabs)
        }
    }
}

File: WoofAndroid/presentation/home/src/main/kotlin/com/perrystreet/woof/presentation/home/viewmodel/HomeViewModel.kt
package com.perrystreet.woof.presentation.home.viewmodel

import com.perrystreet.woof.presentation.common.viewmodel.StateProducingViewModel
import com.perrystreet.woof.presentation.home.uimodel.HomeTabUIModel
import org.koin.android.annotation.KoinViewModel

@KoinViewModel
class HomeViewModel : StateProducingViewModel<HomeViewModel.State>(initialValue = State.Initial) {

    data class State(
        val selectedTab: HomeTabUIModel,
        val tabs: List<HomeTabUIModel>,
    ) {
        val isBackHandled: Boolean
            get() = selectedTab != HomeTabUIModel.Browse

        companion object {
            val Initial = State(selectedTab = HomeTabUIModel.Browse, tabs = HomeTabUIModel.entries)
        }
    }

    fun onTabSelect(tab: HomeTabUIModel) {
        _state.onNext(currentState.copy(selectedTab = tab))
    }

    fun onBackTap() {
        _state.onNext(currentState.copy(selectedTab = HomeTabUIModel.Browse))
    }
}
